File: 0797 - All Paths From Source to Target/All Paths From Source to Target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        visited = len(graph) * [False]
        visited[0] = True
        result = []
        for el in graph[0]:
            self.dfs(graph, result, [0], el, visited)
        return result
        
    def dfs(self, graph: List[List[int]], result: List[List[int]], path: List[int], vertex: int, visited: List[bool]) -> None:
        if vertex == len(graph) - 1:
            path.append(vertex)
            result.append(path)
            return

        if visited[vertex] == True:
            return
        
        visited[vertex] = True
        for el in graph[vertex]:
            self.dfs(graph, result, path + [vertex], el, visited)
        visited[vertex] = False

File: 0797 - All Paths From Source to Target/All Paths From Source to Target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<bool> visited(graph.size(), false);
        vector<vector<int>> result;
        vector<int> path;
        path.push_back(0);
        for (auto el : graph[0]) {
            dfs(result, path, visited, el, graph);
        }
        return result;
    }
    
    void dfs(vector<vector<int>>& result, vector<int>& path, vector<bool>& visited, int vertex, const vector<vector<int>>& graph) {
        if (vertex == visited.size() - 1) {
            path.push_back(vertex);
            result.push_back(path);
            path.pop_back();
        }
        
        if (visited[vertex] == true) {
            return;
        }
        visited[vertex] = true;
        for (auto el : graph[vertex]) {
            path.push_back(vertex);
            dfs(result, path, visited, el, graph);
            path.pop_back();
        }
        visited[vertex] = false;
    }
};
